### PR TITLE
Adapting FastINLA to work for a variety of basket trials including survival analysis.

### PR DESCRIPTION
**Pull request by tbenthompson**
- Extracted the `log_joint` and `grad_hess` functions from the numpy_inference pathway through FastINLA. This allows replacing those functions with other models. These models still need to have the hierarchical sigma2-InvGamma + theta-Normal structure, but can have different outcome distributions. In particular, in the other repo, I have examples using this for survival analysis.
- Modified FastINLA to take a `data` parameter instead of `y` and `n`. This helps abstract the concept of data across different model types. 

~~Currently, the tests are failing because of the JAX-float32 stuff. I don't like the precedent of merging PRs with failing tests, so I'd prefer to either:~~
~~1. Wait until @constinit PR is ready to merge until merging this.~~
~~2. Switch JAX to use float64 for this PR. Then @constinit can revert this change in his PR.~~

~~Fine with me either way.~~

### CODE REVIEW

File: research/berry/berrylib/fast_inla.py [hunk 1]
```diff
@@ -217,30 +277,13 @@ def optimize_mode(self, y, n, fixed_arm_dim=None, fixed_arm_values=None):
...
-
-    def jax_inference(self, y, n):
+    def jax_inference(self, data):
         """
         See the numpy implementation for comments explaining the steps. The
         series of operations is almost identical in the JAX implementation.
         """
-        y = jnp.asarray(y)
-        n = jnp.asarray(n)
+        y = jnp.asarray(data[..., 0], dtype=np.float64)
```
Comment (constinit) [inline research/berry/berrylib/fast_inla.py:285]: Why add float64 here?

File: research/berry/berrylib/test_berry.py [hunk 2]
```diff
@@ -200,8 +202,8 @@ def test_fast_inla(method, N=10, iterations=1):
 
     # Do our comparisons in probability space.
     np.testing.assert_allclose(
-        logistic(theta_max[0, 12]),
```
Comment (constinit) [inline research/berry/berrylib/test_berry.py:203]: This logistic is correct I thought?

File: research/berry/berrylib/fast_inla.py (hunk 1 above)
Comment (tbenthompson) [reply research/berry/berrylib/fast_inla.py:285]: Oh, I can revert this. I was trying to see if the M1 issues we were having went away when I made very certain everything was float64. Those problems were solved with your PR though so we should be all set now. I'll remove this.

File: research/berry/berrylib/test_berry.py (hunk 2 above)
Comment (constinit) [reply research/berry/berrylib/test_berry.py:203]: Fair enough, though do please remove `# Do our comparisons in probability space.`

**Pull request merged**